### PR TITLE
Complete Async

### PR DESCRIPTION
**Pull request by nomisRev**
A couple of things missing from `Async` are added here such as `asyncF`, `cancellable`, `cancellableF` and `shift` which also adds default implementation for `continueOn`.

`cancellable` belongs in `Async` but the laws require `Concurrent` to properly test it. Thoughts? Should I just add some tests using concrete types?

### CODE REVIEW

**Comment by JorgeCastilloPrz** [comment]
Can we test that on Concurrent PR?

File: modules/effects/arrow-effects-kotlinx-coroutines-instances/src/test/kotlin/arrow/effects/DeferredTest.kt [hunk 1]
```diff
@@ -176,5 +178,34 @@ class DeferredKTest : UnitSpec() {
         }
       }
     }
+
+    "test" {
```
Comment (JorgeCastilloPrz) [inline modules/effects/arrow-effects-kotlinx-coroutines-instances/src/test/kotlin/arrow/effects/DeferredTest.kt]: Is this temporary or should we rename this to something more meaningful? :d

File: modules/effects/arrow-effects-instances/src/test/kotlin/arrow/effects/IOTest.kt [hunk 2]
```diff
@@ -374,5 +376,22 @@ class IOTest : UnitSpec() {
       }.fix()
       result.unsafeRunSync() shouldBe 2
     }
+
+    "IO bracket cancellation should release resource with cancel exit status" {
```
Comment (JorgeCastilloPrz) [inline modules/effects/arrow-effects-instances/src/test/kotlin/arrow/effects/IOTest.kt]: :+1:

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt [hunk 3]
```diff
@@ -67,6 +68,15 @@ fun <A> Deferred<A>.k(): DeferredK<A> =
 fun <A> CoroutineScope.asyncK(ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, f: suspend CoroutineScope.() -> A): DeferredK<A> =
   DeferredK.Generated(ctx, start, this) { f() }
 
+fun <A> CoroutineScope.asyncK2(ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, fa: suspend ((Either<Throwable, A>) -> Unit) -> Unit): DeferredK<A> =
```
Comment (JorgeCastilloPrz) [inline modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: Why 2? Can't we have more semantic names somehow related to how this one works?

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt [hunk 4]
```diff
@@ -294,15 +304,32 @@ sealed class DeferredK<A>(
...
+    when (this) {
+      is Generated -> DeferredK.Generated(ctx, coroutineStart, _scope) {
+        _bracketCase(_scope.async(ctx, coroutineStart) { generator() }, use, release)
+      }
+      is Wrapped -> Generated(Dispatchers.Unconfined, CoroutineStart.LAZY, _scope) {
+        _bracketCase(memoized, use, release)
       }
     }
 
+  private suspend fun <B> _bracketCase(self: Deferred<A>, use: (A) -> DeferredK<B>, release: (A, ExitCase<Throwable>) -> DeferredK<Unit>): B {
```
Comment (JorgeCastilloPrz) [inline modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: Are we overall naming private stuff with an underscore preffix ?

File: modules/core/arrow-core/src/test/kotlin/arrow/core/IdTest.kt [hunk 5]
```diff
@@ -25,7 +28,6 @@ class IdTest : UnitSpec() {
       SemigroupLaws.laws(Id.semigroup(Int.semigroup()), Id(1), Id(2), Id(3), Id.eq(Int.eq())),
       MonoidLaws.laws(Id.monoid(Int.monoid()), Id(1), Id.eq(Int.eq())),
       ShowLaws.laws(Id.show(), Eq.any()) { Id(it) },
-      MonoidLaws.laws(Id.monoid(String.monoid()), Id("Hello"), Eq.any()),
```
Comment (pakoito) [inline modules/core/arrow-core/src/test/kotlin/arrow/core/IdTest.kt]: ???

Comment (nomisRev) [reply modules/core/arrow-core/src/test/kotlin/arrow/core/IdTest.kt]: This was overlap between my branch & master. Fixed within merge.

**Comment by raulraja** [comment]
🎉 👏

**Comment by JorgeCastilloPrz** [comment]
Awesome job Simon 👏👏

**Comment by nomisRev** [comment]
Thanks :)

**Pull request merged**